src/ray/object_manager/pull_manager.cc

```diff
@@ -0,0 +1,135 @@
+#include "ray/object_manager/pull_manager.h"
+
+namespace ray {
+
+PullManager::PullManager(
+    NodeID &self_node_id, const std::function<bool(const ObjectID &)> object_is_local,
+    const std::function<void(const ObjectID &, const NodeID &)> send_pull_request,
+    const std::function<int(int)> get_rand_int,
+    const RestoreSpilledObjectCallback restore_spilled_object)
+    : self_node_id_(self_node_id),
+      object_is_local_(object_is_local),
+      send_pull_request_(send_pull_request),
+      get_rand_int_(get_rand_int),
+      restore_spilled_object_(restore_spilled_object) {
+  object_is_local(ObjectID::FromRandom());
+}
+
+bool PullManager::Pull(const ObjectID &object_id, const rpc::Address &owner_address) {
+  RAY_LOG(DEBUG) << "Pull "
+                 << " of object " << object_id;
+  // Check if object is already local.
+  if (object_is_local_(object_id)) {
+    RAY_LOG(ERROR) << object_id << " attempted to pull an object that's already local.";
+    return false;
+  }
+  if (pull_requests_.find(object_id) != pull_requests_.end()) {
+    RAY_LOG(DEBUG) << object_id << " has inflight pull_requests, skipping.";
+    return false;
+  }
+
+  pull_requests_.emplace(object_id, PullRequest());
```

Comment:
Shouldn't we call TryPull() here?

Reply:
I don't think we can do that because we don't have object's locations still since OnLocationChange needs to do `it->second.client_locations = std::vector<NodeID>(client_ids.begin(), client_ids.end());` first

Reply:
Hmmm can you take a look at the return value doc in the header? Lemme know if that needs clarification.